apps/aecore/src/aec_tx_pool_sync.erl

```diff
@@ -0,0 +1,477 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2018, Aeternity Anstalt
+%%% @doc Memory pool synchronization
+%%%
+-module(aec_tx_pool_sync).
+
+-behaviour(gen_server).
+
+%% API for supervisor
+-export([ start_link/0
+        , stop/0
+        ]).
+
+%% Sync API
+-export([ accept/2
+        , connect/2
+        , sync_finish/2
+        , sync_get/2
+        , sync_unfold/2
+        ]).
+
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-import(aec_peers, [ppp/1]).
+
+-record(sync,
+        { id           :: reference()
+        , peer_id      :: binary()
+        , peer_con     :: pid()
+        , peer_con_ref :: undefined | reference()
+        , tree         :: undefined | aeu_mp_trees:tree()
+        , data         :: undefined | [binary()]
+        }).
+
+-type sync_object() :: #sync{}.
+
+-record(state,
+        { local  = not_synced :: not_synced | {active, sync_object()} | in_sync
+        , remote = []         :: [sync_object()]
+        }).
+
+-define(SERVER, ?MODULE).
+-define(MAX_INCOMING_SYNC, 5).
+
+%% -- API --------------------------------------------------------------------
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+stop() ->
+    gen_server:stop(?SERVER).
+
+accept(PeerId, PeerCon) ->
+    gen_server:call(?SERVER, {accept, PeerId, PeerCon}).
+
+connect(PeerId, PeerCon) ->
+    gen_server:cast(?SERVER, {connect, PeerId, PeerCon}).
+
+sync_finish(PeerId, Status) ->
+    gen_server:cast(?SERVER, {finish, PeerId, Status}).
+
+sync_get(PeerId, TxHashes) ->
+    gen_server:call(?SERVER, {get, PeerId, TxHashes}).
+
+sync_unfold(PeerId, Unfolds) ->
+    gen_server:call(?SERVER, {unfold, PeerId, Unfolds}).
+
+%% -- gen_server callbacks ---------------------------------------------------
+
+init([]) ->
+    {ok, #state{}}.
+
+handle_call({accept, PeerId, PeerCon}, _From, State) ->
+    handle_accept(State, PeerId, PeerCon);
+handle_call({get, PeerId, TxHashes}, _From, State) ->
+    Res = handle_get(State, PeerId, TxHashes),
+    {reply, Res, State};
+handle_call({unfold, PeerId, Unfolds}, _From, State) ->
+    Res = handle_unfold(State, PeerId, Unfolds),
+    {reply, Res, State};
+handle_call(Request, From, State) ->
+    lager:warning("Ignoring unknown call request from ~p: ~p", [From, Request]),
+    {noreply, State}.
+
+handle_cast({connect, PeerId, PeerCon}, State) ->
+    {noreply, handle_connect(State, PeerId, PeerCon)};
+handle_cast({finish, PeerId, _Status}, State) ->
+    {noreply, handle_finish(State, PeerId)};
+handle_cast(Msg, State) ->
+    lager:warning("Ignoring unknown cast message: ~p", [Msg]),
+    {noreply, State}.
+
+handle_info({local_action, Ref, TimerRef, Action, Result}, State) ->
+    erlang:cancel_timer(TimerRef),
+    %% We might have already gotten the message :-(
+    receive {timeout, TimerRef, _} -> ok
```

Comment:
Who adds this `{timeout, _}` wrapper tuple?

----

```
$ erl
Erlang/OTP 20 [erts-9.2.1] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V9.2.1  (abort with ^G)
1> Timeout = 1000, TimeoutMsg = {local_action, a_ref, a_pid, an_action}, ok.
ok
2> erlang:send_after(Timeout, self(), TimeoutMsg).
#Ref<0.959383622.920387587.49031>
3> flush().
Shell got {local_action,a_ref,a_pid,an_action}
ok
```

Reply:
`erlang:start_timer/3,4`